nexus/src/app/sagas/silo_create.rs

```diff
@@ -0,0 +1,196 @@
+            saga_create_silo_admin_group,
+            saga_create_silo_admin_group_undo,
+        ),
+    );
+
+    template_builder.build()
+}
+
+async fn saga_create_silo_record(
+    sagactx: ActionContext<SagaSiloCreate>,
+) -> Result<db::model::Silo, ActionError> {
+    let osagactx = sagactx.user_data();
+    let params = sagactx.saga_params();
+
+    let silo_id = sagactx.lookup::<Uuid>("silo_id")?;
+    let opctx = OpContext::for_saga_action(&sagactx, &params.serialized_authn);
+
+    let silo = osagactx
+        .datastore()
+        .silo_create(
+            &opctx,
+            db::model::Silo::new_with_id(silo_id, params.create_params.clone()),
+        )
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    Ok(silo)
+}
+
+async fn saga_create_silo_record_undo(
+    sagactx: ActionContext<SagaSiloCreate>,
+) -> Result<(), anyhow::Error> {
+    let osagactx = sagactx.user_data();
+    let params = sagactx.saga_params();
+    let nexus = osagactx.nexus();
+
+    let opctx = OpContext::for_saga_action(&sagactx, &params.serialized_authn);
+    let silo_id = sagactx.lookup::<Uuid>("silo_id")?;
+
+    nexus.silo_delete_by_id(&opctx, silo_id).await?;
+
+    Ok(())
+}
+
+async fn saga_create_silo_admin_group(
+    sagactx: ActionContext<SagaSiloCreate>,
+) -> Result<Option<db::model::SiloGroup>, ActionError> {
+    let osagactx = sagactx.user_data();
+    let params = sagactx.saga_params();
+
+    if let Some(admin_group_name) = &params.create_params.admin_group_name {
+        let opctx =
+            OpContext::for_saga_action(&sagactx, &params.serialized_authn);
+
+        let silo = sagactx.lookup::<db::model::Silo>("created_silo")?;
+        let silo_admin_group_id =
+            sagactx.lookup::<Uuid>("silo_admin_group_id")?;
+
+        let silo_group = osagactx.datastore().silo_group_create(
+            &opctx,
+            db::model::SiloGroup::new(
+                silo_admin_group_id,
+                IdentityMetadataCreateParams {
+                    name: admin_group_name.parse().map_err(|_|
+                        ActionError::action_failed(
+                            Error::invalid_request(&format!(
+                                "could not parse admin group name {} during silo_create",
+                                admin_group_name,
+                            ))
+                        )
+                    )?,
+                    description: "".into(),
+                },
+                silo.id(),
+            )
+        )
+        .await
+        .map_err(ActionError::action_failed)?;
+
+        // Grant silo admin role for members of the admin group.
+        let policy = shared::Policy {
+            role_assignments: vec![shared::RoleAssignment {
+                identity_type: shared::IdentityType::SiloGroup,
+                identity_id: silo_group.id(),
+                role_name: authz::SiloRole::Admin,
+            }],
+        };
+
+        osagactx
```

Comment:
My (potentially incorrect) understanding is that both `silo_group_create` and `silo_update_policy` are atomic - they either complete or fail (and leave the db untouched). If this is true, then `saga_create_silo_admin_group` can fail at one of those two points. If it fails at `silo_update_policy`, then it has to delete the group. If it fails at `silo_group_create` then there's no cleanup required.

This is only true if there are no additional saga steps though! If my understanding is correct, I could add a comment explaining as much, otherwise I'm open to changing this.

If I do need to change this, I don't see a way for sagas to store arbitrary context using a method like

```rust
sagactx.save::<shared::Policy>("silo_previous_policy", osagactx.datastore().silo_fetch_policy(opctx, &name))?;
```

I'll have to add a separate saga node that only saves the current policy before modification.

Reply:
0aa76698 changed to use a transaction